crates/optimism/txpool/src/maintain.rs

```diff
@@ -75,3 +105,103 @@ where
         }
     }
 }
+
+/// Returns a spawnable future for maintaining the state of the transaction pool.
+pub fn maintain_transaction_pool_interop_future<N, Pool, St>(
+    pool: Pool,
+    events: St,
+    supervisor_client: SupervisorClient,
+) -> BoxFuture<'static, ()>
+where
+    N: NodePrimitives,
+    Pool: TransactionPool + 'static,
+    Pool::Transaction: MaybeInteropTransaction,
+    St: Stream<Item = CanonStateNotification<N>> + Send + Unpin + 'static,
+{
+    async move {
+        maintain_transaction_pool_interop(pool, events, supervisor_client).await;
+    }
+    .boxed()
+}
+
+/// Maintains the state of the transaction pool by handling new blocks and reorgs.
+///
+/// This listens for any new blocks and reorgs and updates the transaction pool's state accordingly
+pub async fn maintain_transaction_pool_interop<N, Pool, St>(
+    pool: Pool,
+    mut events: St,
+    supervisor_client: SupervisorClient,
+) where
+    N: NodePrimitives,
+    Pool: TransactionPool,
+    Pool::Transaction: MaybeInteropTransaction,
+    St: Stream<Item = CanonStateNotification<N>> + Send + Unpin + 'static,
+{
+    let metrics = MaintainPoolInteropMetrics::default();
+    let supervisor_client = Arc::new(supervisor_client);
+    loop {
+        let Some(event) = events.next().await else { break };
+        if let CanonStateNotification::Commit { new } = event {
+            if new.is_empty() {
+                continue;
+            }
+            let timestamp = new.tip().timestamp();
+            let mut to_remove = Vec::new();
+            let mut to_revalidate = Vec::new();
+            for tx in &pool.pooled_transactions() {
+                // Only interop txs have this field set
+                if let Some(interop) = tx.transaction.interop() {
+                    if !interop.is_valid(timestamp) {
+                        // That means tx didn't revalidated during [`OFFSET_TIME`] time
+                        // We could assume that it won't be validated at all and remove it
+                        to_remove.push(*tx.hash());
+                    } else if interop.is_stale(timestamp, OFFSET_TIME) {
+                        // If tx has less then [`OFFSET_TIME`] of valid time we revalidate it
+                        to_revalidate.push(tx.clone())
+                    }
+                }
+            }
```

Comment:
I see, this will always revalidate txs that became stale and try revalidate those, extending their timeout.

this kinda means we could optimize this and instead of doing this on a per block basis, we can re-check based on an interval derived from the offset time.

but for this pr this approach seems okay.

Reply:
Yes, we could make this more intelligent
Would be nice to wait for supervisor to support batches natively so we could make it superb